src/main/java/com/google/codeu/data/Datastore.java

```diff
@@ -155,22 +155,100 @@ public User getUser(String email) {
+   *
+   * @return a Set of users, no order to this list, empty list if there are no users that have posted
+   */
+  public Set<String> getUsers () {
+    Set<String> users = new HashSet<>();
+    Query query = new Query("Message");
+    PreparedQuery results = datastore.prepare(query);
+    for (Entity entity : results.asIterable()) {
+      users.add((String) entity.getProperty("user"));
+    }
+    return users;
+
+  }
+
+  /** Stores a new  Place in Datastore. */
+  public void storePlace(Place place) {
+    Entity placeEntity = new Entity("Place", place.getId().toString());
+    placeEntity.setProperty("owner", place.getOwner());
+    placeEntity.setProperty("title", place.getTitle());
+    placeEntity.setProperty("description", place.getDescription());
+    placeEntity.setProperty("latitude", place.getLatitude());
+    placeEntity.setProperty("longitude", place.getLongitude());
+    placeEntity.setProperty("timestamp", place.getTimestamp());
+
+    datastore.put(placeEntity);
+  }
+
+  /** Get all the messages currently in the Datastore. */
+  public List<Place> getAllPlaces() {
+    List<Place> places = new ArrayList<>();
+    Query query = new Query("Place")
+        .addSort("timestamp", SortDirection.DESCENDING);
+
+    PreparedQuery results = datastore.prepare(query);
+    for (Entity entity : results.asIterable()) {
+      try {
+        String idString = entity.getKey().getName();
+        UUID id = UUID.fromString(idString);
+        String owner = (String) entity.getProperty("owner");
+        String title = (String) entity.getProperty("title");
+        String description = (String) entity.getProperty("description");
+        long latitude = (long) entity.getProperty("latitude");
+        long longitude = (long) entity.getProperty("longitude");
+        long timestamp = (long) entity.getProperty("timestamp");
+        Place place = new Place(id, owner, title, description, latitude, longitude, timestamp);
+        places.add(place);
+      } catch (Exception e) {
+        System.err.println("Error reading message.");
+        System.err.println(entity.toString());
+        e.printStackTrace();
       }
-      return users;
+    }
+    return places;
+  } 
 
+  /**
+   * Get List of locations created by a specific user.
+   *
+   * @return a list of locations posted by the user, or empty list if user has never posted a
+   * location. List is sorted by time descending.
+   */
+  public List<Place> getPlaces(User user) {
+    String userEmail = user.getEmail();
+    List<Place> places = new ArrayList<>();
+
+    Query query = new Query("Place").setFilter(new Query.FilterPredicate("owner", FilterOperator.EQUAL, userEmail))
+        .addSort("timestamp", SortDirection.DESCENDING);
+    PreparedQuery results = datastore.prepare(query);
+
+    for (Entity entity : results.asIterable()) {
+      try {
+        String idString = entity.getKey().getName();
```

Comment:
This code is similar to above; can you deduplicate?

Reply:
I see what you mean! I based this on past code, so I thought it would be good. I'd be happy to fix it once we get other stuff done if that's okay since we're backlogged

Reply:
Please make this change. If you don't know how we can talk in 1:1.

Reply:
done